### PR TITLE
Map `_name` argments to `_id`, look up `run_id` automatically, and output post-processed pandas dfs

### PR DESCRIPTION
**Pull request by rbroc**
Adds several important convenience functionality to make it easier to interact with the API:

Two decorators pre-process arguments:
- `names_to_ids`: Any parameters that end with`_name` are auto-converted to `_id` by looking up the name using the API and matching injecting the ID to the new request.
 For example: `dataset_name=SherlockMerlin` will be replaced with `dataset_id=5`. This behavior is several routes, such as `api.Predictors.get` or `api.Runs.get`. 
- `find_runs`: Look up `run_id` using the following entities: `['dataset_name', 'task_name', 'subject', 'number', 'session']`. This enables users to look up Predictor by the entities of interest, without having to manually look up the `run_ids`. For example, `api.Predictors.get(dataset_name='Life', subject='5')`

The two decorates are applied to `get` routes of each endpoint by setting the following class level attributes to `True`: `_convert_names_to_ids_`, `_find_runs_`.

- Finally, for all `get` routes, the optional keyword argument `return_type` can be set the `df`, and a pandas dataframe will be returned instead of a list of JSON objects. Additionally, all columns ending in `_id` will be autoconverted to their proper name. In simple cases, such as `dataset_id`, a 1:1 mapping will be obtained using the API. In the case of `run_id`, additional columns are added that map onto the run's BIDS entities (e.g. `subject`, `number`, `session`, `acquisition`).

### CODE REVIEW

**Comment by adelavega** [comment]
@rbroc I made a few changes.

I broke it down into two decorators: 
1) The first decorator `dt_name_to_ids` takes `predictor_name` and `task_name` and converts to `dataset_id` and `task_id`
2) The second one takes `predictor_name` and `task_name` and uses it to call `api.runs.get` to get the `run_id` and inject them.

I have made it so that all `get` calls are decorated with `dt_name_to_ids`, so you can always assume you `.get` functions can take names instead of IDs.

As such, `api.runs.get` can be called with `dataset_name` and `task_name` by the second decorator which uses the outputs of this call to inject `run_ids`.

So far I've only decorated `api.predictors.get` with `find_runs` decorator. 

TODO:

- [ ] Add more filters to `find_runs` such as `subject` and `run`
- [ ] Add tests
- [ ] Add to documentation

**Comment by adelavega** [comment]
I added tests and documentation but the documentation could probably be better. 
The main issue is the API arguments are not easy to reference, and its more confusing now with the added syntactic sugar. 

Also it's currently not yet possible to do the following:

`api.predictor_events(dataset_name='Life')` and get all events. 

For some reason combining `predictor_name` (which should be converted to `predictor_id`), with `dataset_name` does not work.
Most likely because the `predictor_id` it finds is contingent on the `dataset_name`. I can fix that by ensuring the helper function does things sequentially (as we do for `task_name`).

File: pyns/endpoints/base.py [hunk 1]
```diff
@@ -42,3 +64,108 @@ def _base_path_(self):
...
+    api = pyns.Neuroscout()
+    for col in df.columns:
+        if col.endswith('_id'):
+            try:
+                endpoint = getattr(api, col.replace('_id', 's'))
+            except:
+                warnings.warn(f"No API endpoint for {col}, could not convert")
+                
+            if col == 'run_id':
+                ents = ['subject', 'session', 'number', 'acquisition']
```
Comment (rbroc) [inline pyns/endpoints/base.py:135]: minor thing, this is not used - so probably remove

File: docs/source/querying.rst [hunk 2]
```diff
@@ -0,0 +1,142 @@
...
+------------------------------------------
+Friendly outputs to pandas DataFrames
+------------------------------------------
+
+You can easily convert any query result to a pandas DataFrame. Simply pass the argument ``output_type='df'`` to the query.
+This is particularly useful for `PredictorEvents`, as the are naturally represented as a pandas DataFrame`.
+
+::
+
+   >>> neuroscout.predictor_events.get(predictor_name='speech', dataset_name='Sherlock_Merlin', task_name='MerlinMovie')
```
Comment (rbroc) [inline docs/source/querying.rst:120]: `output_type='df'` should be passed here?

Comment (adelavega) [reply docs/source/querying.rst:120]: It's an optional argument

File: pyns/endpoints/base.py (hunk 1 above)
Comment (adelavega) [reply pyns/endpoints/base.py:135]: Why are you saying this is not used?

This adds from `run_id` the columns `subject` `session` etc...

Comment (adelavega) [reply pyns/endpoints/base.py:135]: I'm merging, but if you remember let me know

**Comment by jsmentch** [comment]
I tried out the changes and it's working great, much more user-friendly. I made some minor edits and suggestion to the readthedocs page also: https://github.com/neuroscout/pyNS/pull/108

File: pyns/endpoints/base.py (hunk 1 above)
Comment (rbroc) [reply pyns/endpoints/base.py:135]: the variable `ents` is not used, as far as I can see?

**Comment by adelavega** [comment]
Oh sorry I see it now

**Pull request merged**